Generate new network graphs

willcl-ark opened the pull request:
Generate `networkx.random_internet_as_graph`-type graphs using CLI.

e.g.

```
$ warcli graph create n=100 --outfile=random_internet_as_graph_n100.graphml --random
```

willcl-ark commented:
~~@pinheadmz if you review changed some cli commands:~~

~~`warcli network start` -> `warcli network create <from-file [graph_file] | random>`~~

willcl-ark commented:
The [weighted tags](https://github.com/bitcoin-dev-project/warnet/pull/93/commits/a5be661ef4401232e8effd44c273b9eabb700d28#diff-2abae7faf286f7e28bd593237c1475266c5908a17d871c1f52d26019c50e512cR34) mean that there will likely be more nodes with recent version numbers.

We might want to add a more accurate distribution later, or allow users to pass their own distribution function into this

willcl-ark commented:
@pinheadmz @josibake should we also allow node versionc constriants, "e.g >=0.21.1" or just leave as weighted random over all supported tags?

pinheadmz commented:
> should we also allow node versionc constriants, "e.g >=0.21.1" or just leave as weighted random over all supported tags?

bitcoind should be free to make these decisions. I think once we have the p2p dns seeder (which can provide service flags) this will evolve on its own


I dont love removing `start`, would prefer it if `create` is a new command that once executed either:
- also starts the network as if `start` had also been executed
- save the graph file somewhere and tell the user its name so the user can manully call `start`

this latter option i think is preferable but then i also think well need a directory of graph files and a "list" method for them ?

pinheadmz commented on README.md:
```diff
@@ -129,17 +129,17 @@ Once `warnet` is running it can be interacted with using the cli tool `warcli`.
 This is set by default to `--network="warnet"` to simplify default operation.
 
-To start an example warnet, with your venv active and the server running, run the following command to use the default graph and network:
+To start a random warnet, with your venv active and the server running, run the following command to create 20 nodes where each edge has a 30% chance of being connected:
```
add this instead of replace the example graph method?

pinheadmz commented:
> `warcli network start` -> `warcli network create <from-file [graph_file] | random>`

a little confused. you didnt remove/replace the `start` command ?

willcl-ark commented:
> > should we also allow node versionc constriants, "e.g >=0.21.1" or just leave as weighted random over all supported tags?
> 
> bitcoind should be free to make these decisions. I think once we have the p2p dns seeder (which can provide service flags) this will evolve on its own

Whilst the p2p dns seeder will certainly help in making connections, the bitcoind container cannot decide it's own binary version number it's going to run? This PR is about generating a graph file of randomised (or not) node:versions...

My question was about whether we should have `warcli network create random` take a `min_version` param (or similar), in the case that people might not want v16.3 nodes for example

> 
> I dont love removing `start`, would prefer it if `create` is a new command that once executed either:
> 
>     * also starts the network as if `start` had also been executed
> 
>     * save the graph file somewhere and tell the user its name so the user can manully call `start`
> 
> 
> this latter option i think is preferable but then i also think well need a directory of graph files and a "list" method for them ?

Should we simply remove the grammar at this point, if we are not going to use it for everything? Currently we have:

```log
  debug      Debug commands
   ├─   generate-compose  Generate the docker-compose file for a given...
   ├─   update-dns-seed   Update the dns seed database using a <graph_file> on...
  debug-log  Fetch the Bitcoin Core debug log from <node> in [network]
  help       Display help information for the given command.
  messages   Fetch messages sent between <node_a> and <node_b> in <network>
  network    Network commands
   ├─   down   Run 'docker-compose down on a warnet named <--network> (default:...
   ├─   start  Start a warnet with topology loaded from a <graph_file> into...
   ├─   up     Run 'docker-compose up' on a warnet named <--network> (default:...
  rpc        Call bitcoin-cli <method> <params> on <node> in <--network>
  scenarios  Scenario commands
   ├─   active  List running scenarios on <--network> (default=warnet) as...
   ├─   list    List available scenarios in the Warnet Test Framework
   ├─   run     Run <scenario> from the Warnet Test Framework on <--network>...
   ├─   stop    Stop scenario with <pid> from running on <--network>
  stop       Stop warnetd.
```


@josibake felt like `network` was un-intuative here and would prefer `start` and `stop` as top-level. I'd argue that `create` belongs in `network`, and you now think that `start` and `stop` belong in `network` but not `create` :P

I am very happy to go with the majority, but I think we should spec out exactly how we want this to operate so that we can build it just so!

My personal perference would be to retain the sub-commands (but use them in all cases), but happy to be out-voted on this one...

edit: perhaps on a re-read now, I see that you would be _for_ having `create` (as a `network` sub-command?) but just have it generate the graph file? I could live with that!)

pinheadmz commented:
ok yeah we got wires crossed.

- Create a network with random bitcoind versions: great, got it now. yeah weighting is a cool idea here based on current user agent distribution. This is essentially an automatic warnet generation function and mine as well have this feature. Maybe it can be behind a switch where default is weighted random versions but `--allversions=25.0` or something is a flag

- grammar:
I DO LIKE:
`warcli network create random1`
`warlci network start random1`
`warlci network stop random1`
I think we could also add, example:
```
$ warcli network list
  [
    'random1',
    'example_20nodes`,
    'sensitive_relay'
  ]
```

josibake commented:
Just poppin in to say ignore what I was saying earlier, been playing around with the CI more and got a feel for it. I *do* think more detailed help messages could alleviate any "un-intuitiveness"

willcl-ark commented:
OK how about this push? New sub-command `graph`, which will do graph things. Like generate a random graph!

try

`warcli graph generate 20 0.3` to see one spat out to the terminal. or add `--file=/path/to/file` to have it dumped to a file?

willcl-ark commented:
>     * Create a network with random bitcoind versions: great, got it now. yeah weighting is a cool idea here based on current user agent distribution. This is essentially an automatic warnet generation function and mine as well have this feature. Maybe it can be behind a switch where default is weighted random versions but `--allversions=25.0` or something is a flag

Yeah I think we could add an `--allversions` flag very easily!


> 
>     * grammar:
>       I DO LIKE:
>       `warcli network create random1`
>       `warlci network start random1`
>       `warlci network stop random1`
>       I think we could also add, example:
> 

See latest force-push, again, happy to be overruled on this, but perhaps `graph` sub-command makes _even more_ sense (but takes more steps to fire up?)

pinheadmz commented on src/warnet/warnetd.py:
```diff
@@ -245,29 +248,54 @@ def list_running_scenarios(network: str = "warnet") -> Dict[str, int]:
+    graph = create_graph_with_probability(num_nodes, probability)
+    if file:
+        file = Path(file)
+        networkx.write_graphml(graph, file)
```
are there any options for this? the file works but is a little different than im used to seeing thus far which might make manual editing a bit more confusing e.g. 

```
  <key id="d1" for="node" attr.name="bitcoin_config" attr.type="string" />
  <key id="d0" for="node" attr.name="version" attr.type="string" />
...

      <data key="d0">24.1</data>
      <data key="d1" />
```

pinheadmz commented on src/warnet/cli/network.py:
```diff
@@ -14,20 +14,19 @@ def network():
 @click.option("--force", default=False, is_flag=True, type=bool)
 @click.option("--network", default="warnet", show_default=True)
-def start(
+def from_graph(
```
It might be time to get rid of this default graph, especially for a command called "from graph", i think it should require an argument.

If we do include a new directory `src/graphs` with files like `sens_relay.graphml` and `all_released_versions.graphml` etc they are more than just examples they are like standards. Would it be insane to call them "examples" and then have commands
`warcli network list_examples` and
`warcli network from_example sens_relay` ?
This would be IN ADDITION to `warcli network from_graph <whatever>` which would explicitly require an external file provided by the user

pinheadmz commented on src/warnet/cli/network.py:
```diff
@@ -14,20 +14,19 @@ def network():
 ):
     """
-    Start a warnet with topology loaded from a <graph_file> into <--network> (default: "warnet")
+    Create a warnet with topology loaded from a <graph_file> into <--network> (default: "warnet")
```
might need to drop the default network as well. for example just now i ran a warnet then closed it, then tried to start a new warnet from a new graph and got this:
```
Config dir /Users/matthewzipkin/.warnet/warnet/warnet already exists, not overwriting existing warnet without --force
```

Maybe the dir could be something like...
```
~/.warnet/networks/<graphml file name?>
```

pinheadmz commented on src/warnet/utils.py:
```diff
@@ -31,6 +31,8 @@
     "0.16.3",
     "0.15.2",
 ]
+WEIGHTED_TAGS = [tag for index, tag in enumerate(reversed(SUPPORTED_TAGS)) for _ in range(index + 1)]
```
...or you could just list the versions in `SUPPORTED_TAGS` in order of weight?

pinheadmz commented on src/warnet/warnetd.py:
```diff
@@ -245,29 +248,54 @@ def list_running_scenarios(network: str = "warnet") -> Dict[str, int]:
-        except Exception as e:
-            logger.error(f"Exception {e}")
+@jsonrpc.method()
+def update_dns_seeder(graph_file: str, network: str = "warnet") -> str:
```
its a bit tricky to review moved code, did this function change?

willcl-ark commented:
Force-pushed again, ready for review. Updated README, PR description and commits 👍🏼

pinheadmz commented on src/warnet/warnet.py:
```diff
@@ -338,3 +339,14 @@ def write_fork_observer_config(self):
         logger.info(f"Wrote file: {self.fork_observer_config}")
 
+
+def create_graph_with_probability(num_nodes, p, random_version=False):
```
I think this belongs in utils since its not part of the Warnet class

pinheadmz commented on src/warnet/cli/graph.py:
```diff
@@ -11,6 +14,68 @@ def graph():
+    for module in modules:
+        for attr_name in dir(module):
+            if not attr_name.startswith('_') and 'graph' in attr_name:
+                generators[attr_name] = getattr(module, attr_name)
```
I don't think you need getattr here, just the function name? bc if this is just for user facing cli, we dont really need to see python junk:

```
    'graph_atlas': <function graph_atlas at 0x104b17f60>,
    'graph_atlas_g': <function graph_atlas_g at 0x104b30040>,
    'barbell_graph': <function barbell_graph at 0x104b305e0>,
    'circulant_graph': <function circulant_graph at 0x104b309a0>,
(etc)
```

pinheadmz commented:
help?

```
--> warcli graph create triad_graph
Parsed params: {}
Failed to create graph: triad_graph() missing 1 required positional argument: 'triad_name'
--> warcli graph create triad_graph name1
Invalid parameter format: name1
--> warcli graph create triad_graph --triad_name=name1
Usage: warcli graph create [OPTIONS] GRAPH_TYPE [PARAMS]...
Try 'warcli graph create --help' for help.

Error: No such option: --triad_name
```

pinheadmz commented:
help? pt. 2

```
--> warcli graph create bull_graph
Parsed params: {}
Traceback (most recent call last):
  File "/opt/homebrew/bin/warcli", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/src/warnet/cli/graph.py", line 173, in create
    pos = nx.spring_layout(graph)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/networkx/utils/decorators.py", line 766, in func
    return argmap._lazy_compile(__wrapper)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<class 'networkx.utils.decorators.argmap'> compilation 8", line 3, in argmap_spring_layout_5
  File "/opt/homebrew/lib/python3.11/site-packages/networkx/utils/misc.py", line 257, in create_random_state
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
```

josibake commented:
>     import numpy as np
> ModuleNotFoundError: No module named 'numpy'
> ```

huh, this should have been handled as a dependency when install networkx. @willcl-ark might need to refresh the package dependencies, or @pinheadmz maybe you just need to re-run pip install for the branch?

willcl-ark commented:
Right, I had commits from https://github.com/bitcoin-dev-project/warnet/pull/141 in here so that the cleanup is not so error-filled. Looks like removing them causes CI to fail, so I think we should look at that first @pinheadmz

pinheadmz commented on test/test_base.py:
```diff
@@ -119,6 +119,11 @@ def start_server(self):
         self.cleanup()
 
+    def start_network(self, file=None):
+        graph_file_path = (Path(__file__).parent / "data" / file) if file else DEFAULT_TEST_GRAPH
```
why not `file=DEFAULT_TEST_GRAPH`  and `DEFAULT_TEST_GRAPH=v25_x_12.graphml` ?

willcl-ark replied:
Done

pinheadmz commented:
also, now that we got tests cleaning up properly that don't start networks, I think we should start networks in this test ;-) just to ensure that the generated graph files work.

willcl-ark commented:
> also, now that we got tests cleaning up properly that don't start networks, I think we should start networks in this test ;-) just to ensure that the generated graph files work.

Also did this now. Make sense to have a total end-to-end test.

While I was toying with how much work it would be to switch to pytest I did notice too that we have many tests which re-use the same TestBase object and so could be influenced by a previous test...

pinheadmz commented on test/test_graph.py:
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
```
nit, lets call this `graph_test.py`, I think executable tests should all end in `_test.py`

willcl-ark replied:
In the python test frameworks I've had experience test files take the form of `test_.*.py` and functions inside which are "test" functions are named similarly e.g.: `def test_.*():`

Utility/helper functions which don't begin with `test_` won't be run by the test runner.

IMO that makes more sense, but I can be persuaded

pinheadmz replied:
whatever the naming convention is lets stick to it for all tests and make sure that `test_base.py` does not look like a runnable test

pinheadmz commented on docs/graph.md:
````diff
@@ -67,6 +67,30 @@ Edges can be added between the nodes as follows:
 </edge>
 ```
 
+## Creating graphs
````
Should we add these new commands to warcli.md as well? Or are we getting in to maintenence hell where every new rpc requires multiple doc updates

willcl-ark replied:
Let's not define define commands in two places :)

Can markdown import other markdown? If so we could document each subcommand in one file and import it in two places, but otherwise my preference would be not to...

pinheadmz replied:
seems like Click should be able to auto-document. thatd be nice. Oterwise i dunno the best way to organize. maybe warcli.md has most commands but groups like `scenarios` and `graph` we can link to the individual doc pages instead ?

pinheadmz commented on src/warnet/server.py:
```diff
@@ -252,14 +263,14 @@ def from_file(self, graph_file: str, force: bool = False, network: str = "warnet
+        wn.write_docker_compose()
+        # grep: disable-exporters
+        # wn.write_prometheus_config()
+        wn.write_fork_observer_config()
```
could we also move the zone file stuff out of the thread?

pinheadmz commented on src/warnet/server.py:
```diff
@@ -274,6 +285,37 @@ def thread_start(wn):
         threading.Thread(target=lambda: thread_start(wn)).start()
         return f"Starting warnet network named '{network}' with the following parameters:\n{wn}"
 
+    def graph_list(self) -> List[str]:
```
Do these new methods need comments for the help text? Or does that come from cli/main ?

willcl-ark replied:
They do indeed, nice catch. Will add a line for each here.

willcl-ark replied:
Ah sorry I thought this was in cli.... No, click gets help text from cli functions, so they're fine as-is.

pinheadmz commented on src/warnet/utils.py:
```diff
@@ -395,3 +401,129 @@ def version_cmp_ge(version_str, target_str):
+
+
+def create_graph_with_probability(graph_func, params: List, version: str, bitcoin_conf: Optional[str], random_version: bool):
+
```
nit extra \n

pinheadmz commented on test/build_branch_test.py:
```diff
@@ -1,15 +1,12 @@
 base = TestBase()
 base.start_server()
-print(base.warcli(f"network start {graph_file_path}"))
+base.start_network(file=Path(__file__).parent / "data" / "build_v24_test.graphml")
```
I was hoping we could remove import pathlib and this could just be `base.start_network("build_v24_test")`, let test_base handle the rest

willcl-ark replied:
we can

pinheadmz commented:
got this error running the test:

```
Waiting for all tanks to reach 'running': {'total': 5, 'running': 5}
generated graph created network test passed
killing server with pid 20812

Remaining server output:
2023-10-13 10:01:15,342 - warnet - DEBUG - {'jsonrpc': '2.0', 'method': 'status', 'params': {'network': 'warnet_test_62_k9a_y'}, 'id': 6}
2023-10-13 10:01:15,409 - werkzeug - INFO - 127.0.0.1 - - [13/Oct/2023 10:01:15] "POST /api HTTP/1.1" 200 -

killing server with pid 20812
Exception ignored in atexit callback: <bound method TestBase.cleanup of <test_base.TestBase object at 0x103457550>>
Traceback (most recent call last):
  File "/Users/matthewzipkin/Desktop/work/warnet/test/test_base.py", line 70, in cleanup
    self._cleanup_server()
  File "/Users/matthewzipkin/Desktop/work/warnet/test/test_base.py", line 63, in _cleanup_server
    os.killpg(os.getpgid(self.server.pid), signal.SIGTERM)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
ProcessLookupError: [Errno 3] No such process
```

pinheadmz commented on docs/graph.md:
```diff
@@ -67,6 +67,30 @@ Edges can be added between the nodes as follows:
+warcli graph create --version=25.1 navigable_small_world_graph n=25 p=1 q=3 r=2 dim=2
+
+# Create a random internet AS-style graph with random node versions
+warcli graph create --random random_internet_as_graph n=50
```
should also document `outfile` option since generating a graph is kinda useless without it. or a flag to generate a gaph then network start it immedeatly

willcl-ark replied:
Yeah, writing this pull made me remember we had meant to add an RPC (or inner function) which takes a graph _object_ instead of a file. Then we could pass around graph objects from e.g. a front end.

willcl-ark commented:
> got this error running the test:
> 
> ```
> Waiting for all tanks to reach 'running': {'total': 5, 'running': 5}
> generated graph created network test passed
> killing server with pid 20812
> 
> Remaining server output:
> 2023-10-13 10:01:15,342 - warnet - DEBUG - {'jsonrpc': '2.0', 'method': 'status', 'params': {'network': 'warnet_test_62_k9a_y'}, 'id': 6}
> 2023-10-13 10:01:15,409 - werkzeug - INFO - 127.0.0.1 - - [13/Oct/2023 10:01:15] "POST /api HTTP/1.1" 200 -
> 
> killing server with pid 20812
> Exception ignored in atexit callback: <bound method TestBase.cleanup of <test_base.TestBase object at 0x103457550>>
> Traceback (most recent call last):
>   File "/Users/matthewzipkin/Desktop/work/warnet/test/test_base.py", line 70, in cleanup
>     self._cleanup_server()
>   File "/Users/matthewzipkin/Desktop/work/warnet/test/test_base.py", line 63, in _cleanup_server
>     os.killpg(os.getpgid(self.server.pid), signal.SIGTERM)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ProcessLookupError: [Errno 3] No such process
> ```

That's odd. We should handle that in any case, but I don't see the same?

pinheadmz commented:
also saw this at the end of the test. I feel like if the test is succesful it should stop the server using warcli stop and not have to kill PID. although hm. maybe its trying to read log file after its been deleted?

```
Stopping network
Error stopping server: 'NoneType' object has no attribute 'read'
Attempting to cleanup docker network
killing server with pid 207604

```

willcl-ark commented:
> also saw this at the end of the test. I feel like if the test is succesful it should stop the server using warcli stop and not have to kill PID. although hm. maybe its trying to read log file after its been deleted?
> 
> 
> 
> ```
> 
> Stopping network
> 
> Error stopping server: 'NoneType' object has no attribute 'read'
> 
> Attempting to cleanup docker network
> 
> killing server with pid 207604
> 
> 
> 
> ```

Agree, I'll address later this eve

pinheadmz commented:
tried to create a `sudoku_graph` but failed:

```
ModuleNotFoundError: No module named 'scipy'
```
do some of these graphs require certain dependencies? yeesh

pinheadmz commented:
Related to this PR: we have example graph files in three places now:
`src/templates`
`test/data`
`graphs/`

The last one is where Martin went to find a graph file to play with. I propose we remove `example.graphml` entirely and add a `default.graphml` to `graphs/` that is easy (i.e. 12 v25.0 nodes with no branch building) so it can be started quickly by new users

willcl-ark committed:
graph: add generator

delete random AS graphs of size `n` using either default tag (latest) or
random tags from SUPPORTED_TAGS

willcl-ark commented:
@pinheadmz updated this for you

pinheadmz commented:
need to add lxml to requirements

pinheadmz commented on src/warnet/cli/graph.py:
```diff
@@ -0,0 +1,39 @@
+
+
+@graph.command()
+@click.argument("params", nargs=-1, type=str)
```
lets just require `--n=<integer>` right? there's no other params

willcl-ark replied:
Yep agree.

Although... if we ever extend it to other graph types which _do_ have more params, I think this is useful to leave in.

Personally, I feel like it's not much code, which does no harm currently, and I'd prefer to leave it as it's written up now for the future?

Easily persuaded that's not the best course of action here though.

pinheadmz replied:
it's just weird to me to suddenly have an arg without the `--`. And also, I think filename should either be required, or there should be a flag to pipe the output right back in to `network start`. Would anyone need a 100 node graph file printed to the console?

pinheadmz commented on src/warnet/utils.py:
```diff
@@ -408,3 +417,94 @@ def default_bitcoin_conf_args() -> str:
+
+    logger.debug(f"Parsed params: {kwargs}")
+
+    try:
```
IMO most of the try/catch you have in this function should be removed so the error gets thrown back to the RPC caller. Otherwise you get a nice log message but all the CLI shows is "server error"

willcl-ark replied:
Hmmm, this is an _attempt_. I actually don't recall why now I needed it at all. IIRC during some of my testing some graphs were breaking without it. But could have been an artefact of the `nargs` optional cli args thingy (i.e. dodgy input coming from terminal?)

Perhaps if we want to actually return exceptions back to the user via CLI (do we though!?) we should have a refactor PR which moves all try/except into server.py? Each server function wraps function calls with them and can then either returns success, or any bubbled up exception (can also remove our dodgy decorator at the same time?)

pinheadmz replied:
just execute `warcli graph create` with no arguments.

User gets this:
`Error generating graph: Server error`

Logs get all this. Which starts off making sense (missing an arg) but ends up in a really weird place...
```
2023-11-01 17:42:21,811 - utils - ERROR - Failed to create graph: random_internet_as_graph() missing 1 required positional argument: 'n'
2023-11-01 17:42:21,846 - warnet.server - ERROR - unexpected error
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/flask_jsonrpc/site.py", line 135, in handle_dispatch_except
    return self.dispatch(req_json)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/flask_jsonrpc/site.py", line 184, in dispatch
    resp_view = current_app.ensure_sync(view_func)(**params)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/typeguard/__init__.py", line 1033, in wrapper
    retval = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/src/warnet/server.py", line 291, in graph_generate
    nx.write_graphml(graph, bio)
  File "/opt/homebrew/lib/python3.11/site-packages/networkx/utils/decorators.py", line 766, in func
    return argmap._lazy_compile(__wrapper)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<class 'networkx.utils.decorators.argmap'> compilation 9", line 5, in argmap_write_graphml_lxml_5
    import itertools
             ^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/networkx/readwrite/graphml.py", line 176, in write_graphml_lxml
    writer = GraphMLWriterLxml(
             ^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/networkx/readwrite/graphml.py", line 745, in __init__
    self.add_graph_element(graph)
  File "/opt/homebrew/lib/python3.11/site-packages/networkx/readwrite/graphml.py", line 751, in add_graph_element
    if G.is_directed():
       ^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'is_directed'
```

pinheadmz replied:
But I hear ya, we might need an RPC overhaul to clean up some `try`'s. I think we are inconsistent with it.

willcl-ark commented:
> need to add lxml to requirements

```
₿ pip uninstall warnet; and pip freeze | xargs pip uninstall; and pip install -e .
...
₿ warcli graph create n=100 --outfile=random_internet_as_graph_n100.graphml --random
Generated graph written to file: random_internet_as_graph_n100.graphml
```

Works for me without that (tm) 🤷🏼‍♂️

willcl-ark committed:
test: only teardown networks when created

Graph CI won't create a network and errors heavily on teardown.
Add a flag to TestBase which can bypass this.
Also add a thread to read server stdout into test logs.
And tear this thread down in the correct order to not block.

willcl-ark committed:
graph: add simple test to check graph can be generated successfully

willcl-ark commented:
OK ready FR FR

pinheadmz commented:
hm yeah i didnt get the lxml error this time... ooooohhhkkkaaayyyyyy

pinheadmz commented on test/test_graph.py:
```diff
@@ -0,0 +1,30 @@
+    # Couldn't get this to work with CI CLI...
+    base.rpc("graph_generate", {"params": ["n=10"], "outfile": tf, "random": False, "version": DEFAULT_TAG})
+    fp = Path(tf)
+    if fp.exists():
```
why not use `assert` here ?

willcl-ark replied:
No reason. Changed.

pinheadmz commented on test/test_graph.py:
```diff
@@ -0,0 +1,30 @@
+with tempfile.TemporaryDirectory() as dir:
+    tf = f"{dir}/{str(uuid.uuid4())}.graphml"
+    print(f"Writing test graph to {tf}")
+    # Couldn't get this to work with CI CLI...
```
```diff
diff --git a/test/test_graph.py b/test/test_graph.py
index d397323..0f2d845 100755
--- a/test/test_graph.py
+++ b/test/test_graph.py
@@ -1,7 +1,6 @@
 #!/usr/bin/env python3
 import tempfile
 import uuid
-import sys
 from pathlib import Path
 from test_base import TestBase
 from warnet.utils import DEFAULT_TAG
@@ -13,17 +12,11 @@ base.start_server()
 with tempfile.TemporaryDirectory() as dir:
     tf = f"{dir}/{str(uuid.uuid4())}.graphml"
     print(f"Writing test graph to {tf}")
-    # Couldn't get this to work with CI CLI...
-    base.rpc("graph_generate", {"params": ["n=10"], "outfile": tf, "random": False, "version": DEFAULT_TAG})
-    fp = Path(tf)
-    if fp.exists():
-        print("graph was successfully created.")
-    else:
-        print("graph was not created.")
-        sys.exit(1)
+    print(base.warcli(f"graph create n=10 --outfile={tf} --version={DEFAULT_TAG}", network=False))
+    base.wait_for_predicate(lambda: Path(tf).exists())
 
     # Test that the graph actually works
-    print(base.warcli(f"network start {fp}"))
+    print(base.warcli(f"network start {Path(tf)}"))
     base.wait_for_all_tanks_status(target="running")
     base.warcli("rpc 0 getblockcount")
 
```

willcl-ark replied:
I'll take it

pinheadmz commented:
YES.
